src/game.cpp

```diff
@@ -13210,11 +13230,11 @@ void game::shift_monsters( const int shiftx, const int shifty, const int shiftz
         if( shiftx != 0 || shifty != 0 ) {
             critter.shift( shiftx, shifty );
         }
-        // TODO: with z-levels, this can be removed, instead shift the critter
-        // along shiftz, too. Than make the 3D-inbounds check.
-        if( shiftz == 0 && m.inbounds( critter.posx(), critter.posy() ) ) {
+
+        if( m.has_zlevels() && m.inbounds( critter.pos() ) ) {
```

Comment:
Looks fine.

A completely unrelated bug:
The scrolling combat text does not consider z-levels at all.  If a zombies is exactly above my character and attacks it, the SCT only considers x,y for the direction and ends up storing `CENTER` as movement direction.

This results in an infinite loop in `game::get_player_input`. Calling `advanceStepOffset` there is supposed to change the position of the text, but because the direction along which the text should move is `CENTER`, it does not move at all.

So in short: the SCT overlays creature, `advanceStepOffset` is called to move it away, but the position is not affects by this, so the same creature still interferes with the text.

I assume the onscreen-text should never have the direction `CENTER` or `CENTER_ABOVE` or `CENTER_BELOW`, similar `iDirX` and `iDirY` should not both be 0 (that's actually more important than the direction).